language/vm/vm_runtime/src/interpreter.rs

```diff
@@ -0,0 +1,1208 @@
+            let exit_code = self
+                .execute_code_unit(&mut current_frame, code)
+                .or_else(|err| Err(self.maybe_core_dump(err, &current_frame)))?;
+            match exit_code {
+                ExitCode::Return => {
+                    // TODO: assert consistency of current frame: stack height correct
+                    if create_account_marker == self.call_stack.0.len() {
+                        return Ok(());
+                    }
+                    if let Some(frame) = self.call_stack.pop() {
+                        current_frame = frame;
+                    } else {
+                        return Err(self.unreachable("call stack cannot be empty", &current_frame));
+                    }
+                }
+                ExitCode::Call(idx) => {
+                    let opt_frame = self
+                        .make_call_frame(current_frame.module(), idx)
+                        .or_else(|err| Err(self.maybe_core_dump(err, &current_frame)))?;
+                    if let Some(frame) = opt_frame {
+                        if !self.call_stack.can_push() {
+                            let err = VMStatus::new(StatusCode::CALL_STACK_OVERFLOW);
+                            return Err(self.maybe_core_dump(err, &current_frame));
+                        }
+                        self.call_stack.push(current_frame);
+                        current_frame = frame;
+                    }
+                }
+                ExitCode::CreateAccount => {
+                    // TODO: this code will be removed but at the moment it re-enters execute_main.
+                    // That creates some issue with errors and core dumps reporting which are
+                    // not completely and correctly sorted out to keep the logic manageable
+                    // and given this is going away soon
+                    if !self.call_stack.can_push() {
+                        let err = VMStatus::new(StatusCode::CALL_STACK_OVERFLOW);
+                        return Err(err);
+                    }
+                    self.call_stack.push(current_frame);
+                    self.create_account_opcode()?;
+                    if let Some(frame) = self.call_stack.pop() {
+                        current_frame = frame;
+                    } else {
+                        return Err(VMStatus::new(StatusCode::UNREACHABLE).with_message(
+                            "returning from CreateAccount opcode with no call stack".to_string(),
+                        ));
+                    }
+                }
+            }
+        }
+    }
+
+    /// Execute a MOVE function until a return or a call opcode is found.
+    #[allow(clippy::cognitive_complexity)]
+    fn execute_code_unit(
+        &mut self,
+        frame: &mut Frame<'txn, FunctionRef<'txn>>,
+        code: &[Bytecode],
+    ) -> VMResult<ExitCode> {
+        // TODO: re-enbale this once gas metering is sorted out
+        //let code = frame.code_definition();
```

Comment:
this is more of an issue with the way cost synthesis uses the interpreter. Because it passes random single instructions we need a function that takes a &[Bytecode] but once that is removed (hopefully should be removed soon) we can go back to the commented code which is much more reasonable and expected.
Hope it makes sense...

Reply:
could this be refactored?

Reply:
what you mean refactor? that code is being worked on between @tzakian and myself so in that respect there is work happening soon. 
But I am not sure what you mean with refactoring here, in other words if there is something immediate you see happening or what